File: rotateArray.cpp
#include <iostream>
using namespace std;

void rotate(int *arr, int n, int k){

    int start = 0;
    int end = k;
    //first reverse arr[0] --> arr[k]
    for(int i=0; i<=k; i++){
        if (start <=end){
            int temp = arr[start];
            arr[start]=arr[end];
            arr[end] = temp;
        }
        start++;
        end--;
    }

    //second reverse arr[n-k] --> arr[n]
    int start1 = n-k;
    int end1 = n-1;

    for(int i=n-k; i<n; i++){
        if (start1 <=end1){
            int temp = arr[start1];
            arr[start1]=arr[end1];
            arr[end1] = temp;
        }
        start1++;
        end1--;
    }

    //third reverse arr[0] --> arr[n]
    int start2 = 0;
    int end2 = n-1;
    for(int i=0; i<n; i++){
        if (start2 <=end2){
            int temp = arr[start2];
            arr[start2]=arr[end2];
            arr[end2] = temp;
        }
        start2++;
        end2--;
    }
}

int main(){
    int k;
    cin >> k;

    int arr[7] = {1,2,3,4,5,6,7};
    int length = sizeof(arr) / sizeof(*arr);
    rotate(arr, length, k);


    cout << "Rotated array by kth place: " ;
    for(int i=0;i<length;i++){
        cout << arr[i] << " ";
    }
    return 0;
}

File: Array/rotateArray.java
package IntroToJava;

public class rotateArray {

    public static void main(String[] args) {

        int[] array = { 10, 20, 30, 40, 50 };

        System.out.println("----------Previous Array---------");
        for(int val : array){
            System.out.print(val + " ");
        }
        System.out.println();
        rotate(array, -3);
    }

    public static void rotate(int[] arr, int k) {
        int n = arr.length;
        k = k % n;

        if (k < 0) {
            k = k + n;
        }

        for( int i=0; i<k; i++){
            int temp = arr[n-1];

            for(int j=n-1; j>0; j--){

                arr[j] = arr[j-1];
            }
            arr[0]= temp;
        }

        System.out.println("----------Rotated Array----------");
        for(int val : arr){
            System.out.print(val + " ");
        }
    }
}
